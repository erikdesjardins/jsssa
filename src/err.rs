use std::fmt::{self, Debug, Display};

pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;

/// Delegates `Debug` to `Display`, since `Termination` always uses `Debug` to print errors
pub struct NiceError(Error);

impl Debug for NiceError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl<T: Into<Error>> From<T> for NiceError {
    fn from(display: T) -> Self {
        NiceError(display.into())
    }
}
