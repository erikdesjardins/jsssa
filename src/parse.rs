use std::env;
use std::fs::File;
use std::process::{Command, Stdio};
use std::io::Write;

use failure::{Error, ResultExt};
use rand::{self, Rng};
use serde_json;

use ast;

static BABYLON_BIN: &[u8] = include_bytes!("../vendor/babylon.js");

pub fn parse(js: &str) -> Result<ast::File, Error> {
    let mut path = env::temp_dir();
    let random_name: String = rand::thread_rng().gen_ascii_chars().take(32).collect();
    path.push(random_name);

    File::create(&path)?
        .write_all(BABYLON_BIN)?;

    let mut child = Command::new("node")
        .arg(&path)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    write!(child.stdin.as_mut().unwrap(), "{}", js)?;

    let out = child.wait_with_output()?;

    if !out.status.success() {
        let stderr = String::from_utf8(out.stderr)?;
        // skip error throwing header
        let stderr = stderr
            .lines()
            .skip(3)
            .collect::<Vec<_>>()
            .join("\n");

        return Err(ParseError { stderr })?;
    }

    let stdout = String::from_utf8(out.stdout)?;
    let ast = serde_json::from_str(&stdout).context("failed to parse babylon output")?;

    Ok(ast)
}

#[derive(Fail, Debug)]
#[fail(display = "JS parse error: {}", stderr)]
struct ParseError {
    stderr: String
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basic_foo() {
        use ast::*;

        assert_eq!(
            parse(
                r#"
                    'use strict';
                    function foo(x) {
                        return x + 1;
                    }
                "#
            ).unwrap(),
            File::new(Program::new(
                vec![
                    FunctionDeclaration::new(
                        Identifier::new("foo".to_owned()),
                        vec![Identifier::new("x".to_owned()).into()],
                        BlockStatement::new(
                            vec![
                                ReturnStatement::new(Some(
                                    BinaryExpression::new(
                                        BinaryOperator::Add,
                                        Box::new(Identifier::new("x".to_owned()).into()),
                                        Box::new(NumericLiteral::new(1.0).into()),
                                    ).into(),
                                )).into(),
                            ],
                            vec![],
                        ),
                        false,
                        false,
                    ).into(),
                ],
                vec![
                    Directive::new(DirectiveLiteral::new("use strict".to_owned())),
                ],
                SourceType::Script,
            ))
        )
    }
}
