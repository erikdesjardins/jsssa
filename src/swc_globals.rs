/// A token proving that SWC's globals are initialized.
pub struct Initialized(());

/// Initializes the globals used by SWC and provides a token as proof.
///
/// A higher-rank lifetime ensures the token cannot escape the closure.
pub fn with<R>(f: impl FnOnce(&'_ Initialized) -> R) -> R {
    swc_common::GLOBALS.set(&swc_common::Globals::new(), || f(&Initialized(())))
}
