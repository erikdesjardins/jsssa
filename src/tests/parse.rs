use crate::err::NiceError;
use crate::parse;
use crate::swc_globals;

macro_rules! case {
    ( $name:ident, $string:expr ) => {
        #[test]
        fn $name() -> Result<(), NiceError> {
            swc_globals::with(|g| {
                let (ast, _) = parse::parse(g, $string)?;
                insta::assert_debug_snapshot!(stringify!($name), ast);
                Ok(())
            })
        }
    };
}

case!(
    basic,
    r#"
    function f(x) {
        while (true);
        x = y.bar;
        z.foo = x ? true : 'hi';
        return +[1 || x, { x }, f + 1, ++g];
    }
    f(1), true;
"#
);

#[test]
fn parse_error() {
    // swc successfully parses `var ab` and returns that AST, along with emitting an error
    swc_globals::with(|g| match parse::parse(g, "var ab-cd = 1;").err() {
        Some(err) => insta::assert_display_snapshot!("parse error", err),
        None => panic!("parse unexpectedly succeeded"),
    });
}

#[test]
fn parse_error_immediate() {
    // swc can't parse anything, making this different from `parse_error`
    swc_globals::with(|g| match parse::parse(g, "*").err() {
        Some(err) => insta::assert_display_snapshot!("parse error immediate", err),
        None => panic!("parse unexpectedly succeeded"),
    });
}

case!(
    string_has_escape_behavior,
    r#"
    "foo";
    "ba\r";
    "ba\\z";
"#
);
