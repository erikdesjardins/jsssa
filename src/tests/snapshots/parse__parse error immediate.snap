---
source: src/tests/parse.rs
expression: err
---
Parse error: TS1109
 --> <anon>:1:1
  |
1 | *
  | ^

error: Unexpected token None
 --> <anon>:1:1
  |
1 | *
  | ^


